src/coordinator/ts/m3db/block_iteration.go

```diff
@@ -145,48 +150,89 @@ func (c *consolidatedSeriesBlockIter) Next() bool {
 		if !nsBlock.Next() {
 			return false
 		}
-		c.index++
 	}
 
 	return true
 }
 
+func (c *consolidatedNSBlockIter) BadNext() bool {
+	// lastDP, _, _ := c.consolidatedNSBlockSeriesIters[c.seriesIndex].Current()
+	lastDP := c.lastDP
+	// c.lastDP = ts.Datapoint{Value: math.NaN()}
+	c.indexTime = c.indexTime.Add(c.bounds.StepSize)
+
+	fmt.Println("index time: ", c.indexTime)
+
+	if c.indexTime.After(c.bounds.End) {
+		return false
+	}
+
+	for c.indexTime.After(lastDP.Timestamp) && c.next() {
+		lastDP, _, _ = c.consolidatedNSBlockSeriesIters[c.seriesIndex].Current()
+		// c.indexTime = c.indexTime.Add(c.bounds.StepSize)
+		c.lastDP = lastDP
+		// return true
+	}
+
+	if c.indexTime.After(lastDP.Timestamp) {
+		fmt.Println("AFTER")
+		// c.indexTime = c.indexTime.Add(c.bounds.StepSize)
+		c.lastDP = ts.Datapoint{Value: math.NaN(), Timestamp: c.indexTime}
+		// return true
+	}
+	c.indexTime = c.indexTime.Add(c.bounds.StepSize)
+
+	fmt.Println("last dp: ", lastDP)
+
+	return true
+}
+
 func (c *consolidatedNSBlockIter) Next() bool {
-	if len(c.consolidatesNSBlockSeriesIters) == 0 {
+	c.indexTime = c.indexTime.Add(c.bounds.StepSize)
+	fmt.Println("index time: ", c.indexTime)
+	lastDP := c.lastDP
+
+	if !c.indexTime.Before(c.bounds.End) {
 		return false
 	}
 
-	c.currentTime = c.currentTime.Add(c.bounds.StepSize)
-	return !c.currentTime.After(c.bounds.End)
+	for c.indexTime.After(lastDP.Timestamp) && c.next() {
+		lastDP, _, _ = c.consolidatedNSBlockSeriesIters[c.seriesIndex].Current()
+		c.lastDP = lastDP
+	}
+	// if c.indexTime.Add(c.bounds.StepSize).Before(lastDP.Timestamp) {
+	// 	c.lastDP = ts.Datapoint{Value: math.NaN(), Timestamp: c.indexTime}
+	// }
+
+	return true
 }
 
 func (c *consolidatedNSBlockIter) next() bool {
 	// todo(braskin): check bounds as well
-	if len(c.consolidatesNSBlockSeriesIters) == 0 {
+	if len(c.consolidatedNSBlockSeriesIters) == 0 {
 		return false
 	}
 
-	for c.seriesIndex < len(c.consolidatesNSBlockSeriesIters) {
-		if c.consolidatesNSBlockSeriesIters[c.seriesIndex].Next() {
+	for c.seriesIndex < len(c.consolidatedNSBlockSeriesIters) {
+		if c.consolidatedNSBlockSeriesIters[c.seriesIndex].Next() {
 			return true
 		}
+		fmt.Println("seriesIndex: ", c.seriesIndex, "indexTime: ", c.indexTime, "bounds: ", c.bounds)
+		// curr, _, _ := c.consolidatedNSBlockSeriesIters[c.seriesIndex].Current()
+		// fmt.Println("current: ", curr)
 		c.seriesIndex++
 	}
-	c.lastVal = true
+	// c.lastVal = true
 
 	return false
 }
 
 func (c *consolidatedNSBlockIter) Current() float64 {
-	lastDP := c.lastDP
+	if !c.indexTime.After(c.lastDP.Timestamp) && c.indexTime.Add(c.bounds.StepSize).After(c.lastDP.Timestamp) {
```

Comment:
nit: Might read better if you pull out `c.indexTime` and `c.lastDP.Timestamp` as variables first